### PR TITLE
DepthFormat + DepthPacking on MeshDepthMaterial

### PR DESCRIPTION
**Pull request by bhouston**
I've redone the MeshDepthMaterial to have two new parameters that allow it to be fully compatible with RGBA packing, Color buffer visualization and highly accurate float buffer output.  It is also able to be compatible with the automatically created z-buffer so that as we adopt using the DEPTH TEXTURE FBO, we do not have to change much.

.depthFormat
- AutoDepthFormat - uses gl_FragCoord.z, which changes based on the camera type.
- LinearClipZDepthFormat - orthographic camera format
- InvClipZDepthFormat - perspective camera format

.depthPacking
- LinearDepthPacking - for color visualization and float buffers
- RGBADepthPacking - for using RGBA buffers to their full extent.

I've checked the math pretty thoroughly.  I've also stayed with the convention that the camera is looking down the negative z-axis.  Thus reconstructed viewZ values are negative.   We often flip around the view space coordinates, e.g. vViewPosition = -mvPosition.xyz, but this isn't correct as it is against OpenGL conventions.

I've been able to remove completely the old depthRGBA that was used by shadow maps and replaced it with MeshDepthMaterial while keeping exactly the same behavior.

### CODE REVIEW

**Comment by bhouston** [comment]
Demo link:

http://ci.threejs.org/api/pullrequests/8573/examples/webgl_materials_channels.html

To prove these transforms are correct you can compare:

Camera: perspective, Channel: depthInvClipZRGBA is the same as Channel: clipZRGBA

Camera: ortho, Channel: depthLinearClipZRGBA is the same as Channel: clipZRGBA

**Comment by mrdoob** [comment]
Oh, seems like my code clean up messed up things. I'll undo these commits.

**Comment by bhouston** [comment]
Just made another improvement, was able to get rid of depthRGBA shader used by the shadowmaps, Shadows now just use MeshDepthMaterial with AutoDepthFormat and RGBADepthPacking. :)  Thus this really does unify together the various depth buffers and it is also future proof. :)

**Comment by mrdoob** [comment]
Awesome!

**Comment by mrdoob** [comment]
While you're at it. Do you mind doing a spaces to tabs in `ClipZRGBAUnpackedShader.js`?

**Comment by bhouston** [comment]
I just removed it, I ended up not using it. :)

**Comment by bhouston** [comment]
hold off for a sec, I've got a bug somewhere with the shadows.

**Comment by mrdoob** [comment]
@WestLangley Looks good?

**Comment by bhouston** [comment]
Got the bug.

File: examples/webgl_materials_channels.html [hunk 1]
```diff
@@ -51,7 +51,7 @@
 
 
 		<script src="js/loaders/BinaryLoader.js"></script>
-		<script src="js/shaders/DepthRGBAUnpackedShader.js"></script>
+		<script src="js/shaders/ClipZRGBAUnpackedShader.js"></script>
```
Comment (mrdoob) [inline examples/webgl_materials_channels.html]: I guess this isn't needed then?

File: examples/webgl_materials_channels.html [hunk 2]
```diff
@@ -169,9 +169,32 @@
...
+				materialDepthInvClipZ = new THREE.MeshDepthMaterial();
+				materialDepthInvClipZ.depthFormat = THREE.InvClipZDepthFormat;
+				materialDepthInvClipZ.depthPacking = THREE.LinearDepthPacking;
+
+				materialDepthInvClipZRGBA = new THREE.MeshDepthMaterial();
+				materialDepthInvClipZRGBA.depthFormat = THREE.InvClipZDepthFormat;
+				materialDepthInvClipZRGBA.depthPacking = THREE.RGBADepthPacking;
+
+				materialClipZRGBA = new THREE.ShaderMaterial( THREE.ShaderLib.clipZRGBA );
+				materialClipZRGBAUnpacked = new THREE.ShaderMaterial( THREE.ClipZRGBAUnpackedShader );
```
Comment (mrdoob) [inline examples/webgl_materials_channels.html]: Nor this?

File: examples/webgl_materials_channels.html (hunk 1 above)
Comment (bhouston) [inline examples/webgl_materials_channels.html]: You are correct, removed. :)

File: examples/webgl_materials_channels.html (hunk 2 above)
Comment (bhouston) [inline examples/webgl_materials_channels.html]: You are correct, removed. :)

**Comment by mrdoob** [comment]
Thanks!

**Comment by bhouston** [comment]
Thank you for being open to the improvements. :)

**Comment by WestLangley** [comment]
@bhouston This is excellent.

I have some questions, however.

Can you explain your motivation for these function names: `viewZToLinearClipZ()` and `viewZToInvClipZ()`? Why "Clip"? Why "InvClip"?

How about `viewDistanceToDepth()` and `viewDistanceToDepthOrtho()`, instead? That would mean passing in a distance, not a coordinate value, to your functions.

My next question is: how did you come up with your mapping formulas?

Finally, your depth maps are dark for near distances; white for far distances. Usually, it is the opposite, I think.

**Comment by bhouston** [comment]
> Can you explain your motivation for these function names: viewZToLinearClipZ() and viewZToInvClipZ()? Why "Clip"? Why "InvClip"?

I think it is linear (orthogonal has a linear z buffer) and inv (perspective has a non-linear z buffer because there is a divide), but clip may be the wrong term because I think it has an extra transform to move the clip, which is usually in the [-1,1] range to the [0,1] range required for the depth buffer.

> How about viewDistanceToDepth() and viewDistanceToDepthOrtho(), instead?

That can work.  I'd prefer viewZToPerspectiveDepth() and viewZToOrthoDepth() but I think it is nearly the same.

> My next question is: how did you come up with your mapping formulas?

They are derived from the orthographic and perspective matrices combined with the transform of clipZ range of [-1,1] to the [0,1] range.  This is why they match gl_FragCoord.z (e.g. AutoDepthFormat), and because they match gl_FragCood.z exactly, you know they are right.

The other equations are just inversions of back from gl_FragCoord.z to view.Z, so we can unpack the depth values properly.

> Finally, your depth maps are dark for near distances; white for far distances. Usually, it is the opposite, I think.

Again, they are based on what is in the true depth buffer and the match gl_FragCoord.z which if we use a Depth Buffer FBO, it is what is in there.  I have used this convention so that as we build upon it we can use @mattdesl's new PR for true depth buffers and we have to change nothing.

**Comment by mrdoob** [comment]
> That can work. I'd prefer viewZToPerspectiveDepth() and viewZToOrthoDepth() but I think it is nearly the same.

These sound good to me too :+1:

**Comment by WestLangley** [comment]
@bhouston 

ref: http://www.songho.ca/opengl/gl_projectionmatrix.html

>GL_PROJECTION matrix is used for this projection transformation. First, it transforms all vertex data from the eye coordinates to the clip coordinates. Then, these clip coordinates are also transformed to the normalized device coordinates (NDC) by dividing with w component of the clip coordinates.

You are mapping from eye space to clip space, then to NDC space in [ -1, 1], and then you are remapping to [0, 1].

So using the nomenclature "Clip" was confusing, IMHO. I am happy with `viewZToPerspectiveDepth()` and `viewZToOrthoDepth()`.

**Comment by WestLangley** [comment]
Can we come up with better constant names? I am not sure what to suggest, though...
```javascript
// Depth format
------------------
THREE.AutoDepthFormat = 3100;			// switches based on camera type, uses gl_FragCoord.z
THREE.LinearClipZDepthFormat = 3101;	// used by orthographic cameras
THREE.InvClipZDepthFormat = 3102;		// used by perspective cameras

// Depth packing strategies
------------------

THREE.LinearDepthPacking = 3200;		// for writing to float textures for high precision or for visualizing results in RGB buffers
THREE.RGBADepthPacking = 3201;			// for packing into RGBA buffers.
```

I do not think the "formats" are formats. The packings seem more like formats to me. The "formats" are describing the formula for the data that is stored.

Also, maybe `LinearDepthPacking` should be `FloatDepthPacking`?

**Comment by bhouston** [comment]
LinearClipZDepthFormat -> OrthoDepthFormula?
InvClipZDepthFormat -> PerspectiveDepthFormula?

I'm fine switching LinearDepthPacking to FloatDepthPacking, they seem like synonyms.  The issue is that FloatDepthPacking isn't always floats, often it is just grayscale packing into 8-bit channels, thus it doesn't have the expected precision of a standard opengl float (which is 32 bits.)

**Comment by WestLangley** [comment]
Just throwing out ideas here... : - )

`MeshDepthMaterial` must specify the metric that is used for depth, and the format that is used to encode it.
```javascript
// material.depthMetric
------------------
THREE.FragCoordMetric = 3100;
THREE.OrthographicMetric = 3101;
THREE.PerspectiveMetric = 3102;

// material.depthPacking
------------------
THREE.LinearDepthPacking = 3200; ???
THREE.RGBADepthPacking = 3201;
```

**Comment by bhouston** [comment]
I would support this:

THREE.AutoDepthMetric = 3100;
THREE.OrthoDepthMetric = 3101;
THREE.PerspectiveDepthMetric = 3102;

I think that auto is better than FragCoord metric because auto does suggest that it switches base on context.

Or we completely remove the ability to specify the metric, and it always uses gl_FragCoord.  This is probably the right behavior in all cases, except for debugging situations like what I've been doing while understanding this area fully.

**Comment by WestLangley** [comment]
>Or we completely remove the ability to specify the metric, and it always uses gl_FragCoord.

Yes. Because it does not make sense if one selects the incorrect metric for the camera being used.

**Pull request merged**